rfcs/0160-portable-service-layer.md

```diff
@@ -0,0 +1,893 @@
+  big chance that it will work with other process managers as well.
+* You may target the lowest common denominator. This is not actually true,
+  because with overrides it is still possible to use any feature of a process
+  manager. Moreover, not all services need to be portable -- low-level services
+  such as `cups` and `syslogd` can still be packaged in a process
+  manager-specific way.
+
+# Alternatives
+[alternatives]: #alternatives
+
+This RFC proposes a new strategy to allow multiple process instances of the
+same service to co-exist. Co-existence is possible by avoiding conflicts --
+making sure that state directories and port numbers do not conflict through
+function parameters.
+
+It is also possible to isolate services by using concepts such as Linux
+namespaces. Then it is not required to avoid certain kinds of conflicts.
+Unfortunately, using these features is restricted to Linux only -- this RFC aims
+to be operating system agnostic.
+
+# Prior art
+[prior-art]: #prior-art
+
+This RFC is not the first approach that combines Nix with a process management
+solution:
+
+* The service deployment chapter (chapter 9) of
+  [Eelco Dolstra's PhD thesis](https://github.com/edolstra/edolstra.github.io/blob/master/pubs/phd-thesis.pdf)
+  describes a very early service deployment approach that uses generated
+  start/stop scripts implemented in `bash`. Because shell scripts are so common,
+  they can be used on a variety of operating systems -- in addition to Linux, it
+  can also be used on Darwin and FreeBSD.
+* [NixOS](https://nixos.org) uses `systemd` as its process management solution.
+  Services can be deployed by generating systemd unit configuration files in a
+  Nix expression. NixOS deploys `systemd` units globally and requires super
+  user privileges to get deployed. Moreover, you cannot use the service layer
+  independently from the rest of NixOS.
+* [nix-darwin](https://github.com/LnL7/nix-darwin) provides various NixOS
+  features on macOS/Darwin, including a service management layer that uses the
+  process manager of macOS: `launchd`.
+* [home-manager](https://github.com/nix-community/home-manager) offers a service
+  layer that works with systemd user services, rather than system services. It
+  contains its own set of service configurations that can be deployed as user
+  services.
+* [system-manager](https://github.com/numtide/system-manager) addresses the
+  limitation of NixOS to use systemd units independently and can deploy many
+  services in the NixOS repository on any Ubuntu-based distribution.
+
+Aside from the first option, the biggest drawback is that these solutions are
+not generic and there is only little reuse possible between service
+configurations.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+The integration with NixOS module system still needs to be worked out in detail.
+Currently, I can think of two scenarios. For this a discussion in the Nix
+community is required.
```

Comment:
This proposal might combine very well with the new `pkgs/by-name` directory introduced with [RFC 140](https://github.com/NixOS/rfcs/pull/140), because it could be extended to support a standard `service.nix` file to specify a standardised service definition for a package, without having to be tied to NixOS.

Reply:
@infinisil can you please elaborate on your comment a bit? I'd like to make sure I understand you clearly. I'm not sure if you mean that this `service.nix` file would be an interface (`options`) without implementation (`config`), or something else... and to what extent the separation would go 

I can imagine a `service.nix` which declares a common set of options for a module and then several other files (`systemd.nix`, `supervisord.nix`, `launchd.nix`, etc...) which define different implementations of these options (maybe even declaring some implementation specific options of their own). I love how this could tie into RFC 140!

Or did you have something else in mind?

Reply:
Pretty much something like that yes. One of the potential future ideas with `pkgs/by-name` was to use it not only for package builds, but also the package-specific modules. So in addition to the current `package.nix` file in each directory, there could be a `module.nix` file or so too. This would then also be automatically discovered in some way and added to NixOS (or as here, more generic process management).

The reason I'm suggesting it here is because I imagine we need some kind of migration from the current NixOS modules to more generic modules, and combining that with a transition to `pkgs/by-name` could work nicely. I think of `pkgs/by-name` as a sort of strictly type-checked part of Nixpkgs, where CI prevents you from e.g. committing `package.nix` files that aren't packages. Similarly it could check that `module.nix` is a module with specific options or so.

Reply:
I love it! Thanks for confirming?

Reply:
> a standardised service definition

In other words, an API?

Reply:
> This would then also be automatically discovered in some way and added to NixOS

We don't even have a module based implementation yet, so it's way too soon to start automating and imposing restrictions. This is a very different problem.
I'm all for grouping files by functionality instead of technical type, but easy on the rules please.



> an API?

An interface.

"Application programming" just muddies the waters. It's 3 syllables either way.

Reply:
> This proposal might combine very well with the new `pkgs/by-name` directory introduced with [RFC 140](https://github.com/NixOS/rfcs/pull/140), because it could be extended to support a standard `service.nix` file to specify a standardised service definition for a package, without having to be tied to NixOS.

There is already at least one project doing exactly this.